Draft TTD API and UI

JohnathanKG opened the pull request:
This PR is a draft for declaring adapters as TTD-compatible, buttons to navigate from the debugger GUI, a python API, and a unit test for the TTD API
Currently it is lacking fully proper documentation in the Python API, and custom icons do not work.

JohnKServices committed:
Add prototype TTD API and UI, add TTD API unit test

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/signed)](https://cla-assistant.io/Vector35/debugger?pullRequest=541) <br/>All committers have signed the CLA.

JohnathanKG committed:
Fix debuggerapi.h formatting

JohnathanKG committed:
Minor ffi.h formatting

xusheng6 commented:
Here are a few quick comments so you can start working on it:

1. We should also add support for "step return reverse". The DbgEng does not offer constant like `DEBUG_STATUS_REVERSE_STEP_OVER` for reverse step return. But you can look at how step return (forward) is implemented, basically it just sends a "gu" command (https://github.com/Vector35/debugger/blob/318e61219ced0815351eea16c6388d3e1372e6f7/core/adapters/dbgengadapter.cpp#L1194-L1201). And to step return reverse, you need to send a "g-u" command
2. For the icons, just use the same ones as the forward operations. We will create new ones soon
3. Actions like "Step Over" are connected to keyboard shortcuts. We should do the same for the newly add actions. You can see how I was doing it here: https://github.com/Vector35/debugger/blob/471917726e646d28390d777902e5e846ef01ec7c/ui/ui.cpp#L452. I think we should add a Shift to the keybindings for the forward actions. For example, step over is F8, then we should make step over reverse shift+F8

JohnKServices committed:
Add step return reverse
Add keybinds for reverse execution (shift + original keybind)
Remove TTD unit test

JohnKServices committed:
Merge remote-tracking branch 'mine/ttd' into ttd

xusheng6 commented:
@JohnathanKG could you please remove the TTD related unit test (for now)? I tested and the trace generated on your box does not work properly on mine. And it probably also would not work on our infra.

JohnathanKG committed:
Remove reverse debugging unit test

JohnathanKG committed:
Delete test/binaries/Windows-x86/helloworld.exe.run

JohnathanKG committed:
Delete test/binaries/Windows-x86_64/helloworld.exe.run

xusheng6 commented:
Merged into https://github.com/Vector35/debugger/commit/7cd18122ad881e17a64d9ef63ceb2c56899e21b6. Thanks for the contribution!
